### PR TITLE
Design Guidelines Edits and Question

### PR DESCRIPTION
**Pull request by bryevdv**
This PR make some very minor edits to the Design Guidelines, mostly to add helpful links to outside tools or projects that are mentioned. 

Additionally, a PR seemed like a convenient place to leave some inline questions/comments about some of the other bullet points. 

cc @johanste

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -6,10 +6,10 @@ The Azure SDK delivers a platform for developers to leverage the wide variety of
...
 ## Azure SDK Design Guidelines
 
-- [.NET](https://azuresdkspecs.z5.web.core.windows.net/DotNetSpec.html) ([template](https://github.com/Azure/azure-sdk-for-net/tree/master/src/SDKs/Template/data-plane))
-- [Java](https://azuresdkspecs.z5.web.core.windows.net/JavaSpec.html) ([template](https://github.com/Azure/azure-sdk-for-java/tree/master/template/client))
-- [Python](https://azuresdkspecs.z5.web.core.windows.net/PythonSpec.html) ([template](https://github.com/Azure/azure-sdk-for-python/tree/master/azure-template))
-- [TypeScript](https://azuresdkspecs.z5.web.core.windows.net/TypeScriptSpec.html) ([template](https://github.com/Azure/azure-sdk-for-js/tree/master/packages/%40azure/template))
+- [.NET](https://azuresdkspecs.z5.web.core.windows.net/DotNetSpec.html)
+- [Java](https://azuresdkspecs.z5.web.core.windows.net/JavaSpec.html)
+- [Python](https://azuresdkspecs.z5.web.core.windows.net/PythonSpec.html)
+- [TypeScript](https://azuresdkspecs.z5.web.core.windows.net/TypeScriptSpec.html)
```
Comment (bryevdv) [inline README.md]: All of these `[template]` links 404'd. If there are suitable new links to update them with instead of just removing them, please direct me to them.

File: docs/ReviewProcess.md [hunk 2]
```diff
@@ -1,7 +1,7 @@
 We expect all Azure SDKs to go through rigorous APIs reviews similar to those conducted for all .NET APIs.
 It's critical that the review is conducted early enough to allow time for fixes,
 and sometimes significant API redesign based on the review feedback.
-If you have never not participated in Azure SDK API review,
+If you have never participated in Azure SDK API review,
 we recommend that you schedule a pre-review (consulting session) before you start working on the APIs.
```
Comment (bryevdv) [inline docs/ReviewProcess.md]: Who/where does one schedule a consulting session with?

**Comment by bryevdv** [comment]
Other questions below 

https://github.com/Azure/azure-sdk/blob/238bacdcf7f617390ff03176b18d621111408f30/docs/design/python/DesignGuidelines.mdk#L9-L10
 I notice the Pipeline config has steps to install and test with PyPy but I did not see anything about Jython. Where/how is Jython support maintained under test? For curiosity, what is the rationale for supporting Jython (AFAIK major work on it stopped quite a while ago)

**Comment by johanste** [comment]
> I notice the Pipeline config has steps to install and test with PyPy but I did not see anything about Jython. Where/how is Jython support maintained under test? For curiosity, what is the rationale for supporting Jython (AFAIK major work on it stopped quite a while ago)

Historically, we've seen enough usage of jython that we didn't want to ignore it. However, it is a great question - we should re-check the data we have to see if it makes sense to continue to recommend supporting it.

> There is nothing explicit about test layout. From inspecting a few examples the convention seems to be out-of-tree, within each subcomponent. It might be worthwhile to explicitly document expectation here.

Good point. Do you think that pointing to the engineering system [documentation's](https://github.com/Azure/azure-sdk/blob/master/docs/engineering-system/repo-structure.md) is sufficient, or should we duplicate the hierarchy here?

> I agree with this but wanted to note that it seems slightly at odds with earlier general advice "YOU SHOULD consider copying or linking required code into the SDK component in order to avoid taking a dependency..." If this discrepancy is intentional it might be worth noting explicitly.

Again, good point. I think we are aligned with the spirit in the general guidance. The intent is for us (adparch) to control what external dependencies you take (it is not a not a free-for-all). If you are granted permission to take an external dependency, doing so is preferred over vendoring the package. Which dependencies are deemed acceptable varies between different language ecosystems. Taking a dependency on, say `requests` in Python has been much less problematic than a dependency on `Json.NET` in C#/.NET. 

In other words, the order of priority would be:
1) No unnecessary dependencies.
2) If you need to depend on something, come to us and motivate why it should be a common "blessed" dependency.
3) Only vendor the package if you need a different version than the commonly agreed upon common "blessed" dependency. 

I'll try to wordsmith the documentation to better reflect this.

> Are there any existing examples of implicit namespace packages in the codebase to point to? I have mixed but still not-yet-very-informed feelings about them.

https://github.com/Azure/azure-sdk-for-python/wiki/Azure-packaging contains information that I need to move into the design guidelines (or I need to link to the wiki).

File: README.md (hunk 1 above)
Comment (johanste) [reply README.md]: I'm not the madoku expert on the team - @bterlson speaks this much better than I do - but I believe that these links get substituted by the build.

File: docs/ReviewProcess.md (hunk 2 above)
Comment (johanste) [reply docs/ReviewProcess.md]: @adrianhall, this should probably include some more information...

File: README.md (hunk 1 above)
Comment (bryevdv) [reply README.md]: This is a plain Markdown `.md` file, is it also processed along with the `.mdk` files? I can delete the commit, just let me know.

**Comment by bryevdv** [comment]
> Good point. Do you think that pointing to the engineering system documentation's is sufficient, or should we duplicate the hierarchy here?

Ah I did not know about that, I think a link to that would be great, I will push a commit to add it.

File: README.md (hunk 1 above)
Comment (bterlson) [reply README.md]: These links should always point to latest (and I'll also observe they're up for me). Possibly a CI run was in progress, can you check again?

Comment (bryevdv) [reply README.md]: Just to be clear about where I saw 404s, it was from the GitHub-rendered README.md at https://github.com/Azure/azure-sdk I am still seeing 404s from the `(template)` links there or if I try to navigate to one directly, e.g. https://github.com/Azure/azure-sdk-for-net/tree/master/src/SDKs/Template/data-plane It sounds like these are processed and modified for use elsewhere, just for my own reference, where is that?

**Comment by johanste** [comment]
Engineering docs is here: https://github.com/Azure/azure-sdk/blob/master/docs/engineering-system/repo-structure.md

**Comment by bryevdv** [comment]
@johanste I've added a link in https://github.com/Azure/azure-sdk/pull/246/commits/39cf45ca41069b8ddc02354532391a5f1fc7f788 but I'm not sure if that is the best or appropriate format. I've gleaned that there is some sort of post-processing that will happen to some links, but not much about it that I could find at http://madoko.org/reference.html

File: docs/ReviewProcess.md (hunk 2 above)
Comment (adrianhall) [reply docs/ReviewProcess.md]: Check - it's an internal process - I think opening an issue on this repo is the right place to start though.

Comment (bryevdv) [reply docs/ReviewProcess.md]: @adrianhall please see https://github.com/Azure/azure-sdk/pull/246/commits/6c2955d6173427c7061def85ee509a0175f5b0fa

File: README.md (hunk 1 above)
Comment (bterlson) [reply README.md]: @bryevdv yes indeed, those links were broken. I've committed a fix, hopefully the links work now.

Comment (bryevdv) [reply README.md]: Thanks!

File: docs/ReviewProcess.md (hunk 2 above)
Comment (adrianhall) [reply docs/ReviewProcess.md]: Ack - I believe that's good enough for these guidelines.

**Comment by johanste** [comment]
@JonathanGiles / @bterlson, can one of you verify @bryevdv 's link that he added in [`39cf45c`](https://github.com/Azure/azure-sdk/commit/39cf45ca41069b8ddc02354532391a5f1fc7f788)?

File: docs/ReviewProcess.md [hunk 3]
```diff
@@ -1,8 +1,8 @@
 We expect all Azure SDKs to go through rigorous APIs reviews similar to those conducted for all .NET APIs.
 It's critical that the review is conducted early enough to allow time for fixes,
 and sometimes significant API redesign based on the review feedback.
-If you have never not participated in Azure SDK API review,
-we recommend that you schedule a pre-review (consulting session) before you start working on the APIs.
+If you have never participated in Azure SDK API review,
+we recommend that you schedule a pre-review (consulting session) before you start working on the APIs. Open an issue on <https://github.com/Azure/azure-sdk> to start a discussion.
```
Comment (adrianhall) [inline docs/ReviewProcess.md]: This has already been integrated into the ReviewProcess, along with many other adjustments.  It's probably best to revert this at this point.

File: docs/design/python/DesignGuidelines.mdk [hunk 4]
```diff
@@ -410,21 +410,21 @@ Example:
...
 ~ Must {#python-support-other-async}
-consider supporting async frameworks other than asyncio (i.e. Trio)
+consider supporting async frameworks other than asyncio (i.e. [Trio](https://trio.readthedocs.io/en/latest/))
 ~
 
 ## Testing
 
 ~ Must {#python-use-pytest}
-use pytest as the test framework
+use [pytest](https://docs.pytest.org/en/latest/) as the test framework (see [here](docs/design/python/DesignGuidelines.mdk) for the expected respository struture)
```
Comment (adrianhall) [inline docs/design/python/DesignGuidelines.mdk]: Use an internal link to get to the expected repository structure location.  If we are going to link, we should link to the right place.

File: docs/design/python/DesignGuidelines.mdk [hunk 5]
```diff
@@ -410,21 +410,21 @@ Example:
...
 ~
 
 ~ Must {#python-use-pytest-asyncio}
-consider using pytest-asyncio for testing of async code. 
+consider using [pytest-asyncio](https://github.com/pytest-dev/pytest-asyncio) for testing of async code.
```
Comment (adrianhall) [inline docs/design/python/DesignGuidelines.mdk]: Must consider sounds like "should" to me.  Discuss with @johanste - are we requiring it or advising it; adjust as appropriate.

File: docs/design/python/DesignGuidelines.mdk [hunk 6]
```diff
@@ -458,11 +458,11 @@ use [flake8-docstrings](https://gitlab.com/pycqa/flake8-docstrings) to verify do
 ~
 
 ~ Must {#python-use-black}
-use Black for formatting your code.
+use [Black](https://black.readthedocs.io/en/stable/) for formatting your code.
 ~
 
 ~ Must {#python-use-mypy}
-consider using MyPy to statically check your code
+consider using [MyPy](https://mypy.readthedocs.io/en/latest/) to statically check your code
```
Comment (adrianhall) [inline docs/design/python/DesignGuidelines.mdk]: Another "Must consider" case - make it should or a definitive must after discussion with @johanste.

**Comment by johanste** [comment]
@bryevdv, rather than fixing things up in this PR, I created a new PR GH-302 where I addressed most (all?) of the issues you brought up. I also fixed up the "must consider" (leftovers from earlier markdown-only format) and removed references to msrest.

**Comment by adrianhall** [comment]
Superceded by #302 - please refer all further discussion to that PR.
